subsys/dfu/dfu_target/src/dfu_target_modem_delta.c

```diff
@@ -28,68 +23,47 @@ struct modem_delta_header {
 	uint32_t magic;
 };
 
-static int  fd;
-static int  offset;
+static int offset;
 static dfu_target_callback_t callback;
 
-static int get_modem_error(void)
-{
-	int rc;
-	int err = 0;
-	socklen_t len;
-
-	len = sizeof(err);
-	rc = getsockopt(fd, SOL_DFU, SO_DFU_ERROR, &err, &len);
-	if (rc) {
-		LOG_ERR("Unable to fetch modem error, errno %d", errno);
-	}
-
-	return err;
-}
-
 static int apply_modem_delta_upgrade(void)
 {
 	int err;
 
 	LOG_INF("Scheduling modem firmware upgrade at next boot");
 
-	err = setsockopt(fd, SOL_DFU, SO_DFU_APPLY, NULL, 0);
-	if (err < 0) {
-		if (errno == ENOEXEC) {
-			LOG_ERR("SO_DFU_APPLY failed, modem error %d",
-				get_modem_error());
-		} else {
-			LOG_ERR("SO_DFU_APPLY failed, modem error %d", err);
-		}
+	err = nrf_modem_delta_dfu_update();
+	if (err != 0) {
+		LOG_ERR("Modem firmware upgrade scheduling failed, error %d", err);
 	}
+
 	return 0;
 }
 #define SLEEP_TIME 1
 static int delete_banked_modem_delta_fw(void)
 {
 	int err;
-	socklen_t len = sizeof(offset);
 	int timeout = CONFIG_DFU_TARGET_MODEM_TIMEOUT;
 
 	LOG_INF("Deleting firmware image, this can take several minutes");
-	err = setsockopt(fd, SOL_DFU, SO_DFU_BACKUP_DELETE, NULL, 0);
-	if (err < 0) {
-		LOG_ERR("Failed to delete backup, errno %d", errno);
+	err = nrf_modem_delta_dfu_erase();
+	if (err != 0) {
+		LOG_ERR("Failed to delete backup, error %d", err);
 		return -EFAULT;
 	}
+
+	k_sleep(K_SECONDS(SLEEP_TIME));
```

Comment:
I'm curious, did it fail without adding this sleep?

Reply:
It did not fail without, but the `err` returned from `nrf_modem_delta_dfu_offset()` in the first iteration of the loop was 5 (`NRF_MODEM_DELTA_DFU_INVALID_DATA`). So it was logged (printed to screen), but did not abort or stop the dfu.

Reply:
Maybe we should add a comment in the code about this? IMO there is something eerie about having seemingly random sleep statements in your code. The later sleep is right inside an `if erase pending` so that is self explanatory.

Reply:
That seems strange @anhmolt. Are you able to reproduce that behavior in a system test in `libmodem`?

Reply:
Not able to reproduce with ztest in `libmodem`, so will look some more at this.

Reply:
Looks like the erase is taking some extra time to initialize on the modem side because network is prioritized. Therefore, `NRF_MODEM_DELTA_DFU_INVALID_DATA` is reported back when calling offset function too quickly after erase function.

Have removed the extra delay and made so `NRF_MODEM_DELTA_DFU_INVALID_DATA` is treated the same way as `NRF_MODEM_DELTA_DFU_ERASE_PENDING` in the erase loop.